lib/compress/zstd_compress.c

```diff
@@ -1915,17 +2046,23 @@ static size_t ZSTD_resetCCtx_byCopyingCDict(ZSTD_CCtx* cctx,
     }
 
     ZSTD_cwksp_mark_tables_dirty(&cctx->workspace);
+    assert(params.useRowMatchfinder != ZSTD_urm_auto);
 
     /* copy tables */
-    {   size_t const chainSize = (cdict_cParams->strategy == ZSTD_fast) ? 0 : ((size_t)1 << cdict_cParams->chainLog);
+    {   size_t const chainSize = ZSTD_shouldNotAllocateChainTable(cdict_cParams->strategy, ZSTD_urm_disableRowMatchfinder, /* forCCtx */ 0)
+                                                            ? 0
+                                                            : ((size_t)1 << cdict_cParams->chainLog);
         size_t const hSize =  (size_t)1 << cdict_cParams->hashLog;
 
         ZSTD_memcpy(cctx->blockState.matchState.hashTable,
                cdict->matchState.hashTable,
                hSize * sizeof(U32));
-        ZSTD_memcpy(cctx->blockState.matchState.chainTable,
+        /* Do not copy cdict's chainTable if cctx has parameters such that it would not use chainTable */
+        if (!ZSTD_shouldNotAllocateChainTable(cctx->appliedParams.cParams.strategy, cctx->appliedParams.useRowMatchfinder, /* forCCtx */ 1)) {
```

Comment:
This is a case where negation gets complicated. It has much lower mental overhead to avoid double negations.